### PR TITLE
[DNM until after 2.8 release] Controller godocs

### PR DESCRIPTION
**Pull request by a-blender**
## Problem

 
Add and update Rancher controllers to have detailed `godoc` descriptions. Huge undertaking and will be chunked out into separate PRs; I'd like review on this first one.

Team 2 controllers
* agentupgrade
* aks
* statscontroller

Team 1 / Neo controllers (we need review on from Team 1)
* auth v1 management, prtb, crtb, user and token

_Note: In future PRs, I'll only be adding godocs for Rancher controllers that fall under the responsibility of team 2. This includes the CAPR, management node driver, management agent, management user, and v2 prov controllers. @pgonin I believe the Highlander team should create a separate GH issue for adding godocs to the Hosted provider (AKS, EKS, and GKE) controllers. Reach out to me with any questions._

### Automated Testing

* Test types added/modified:
    * Unit
    * Integration (Go Framework)
    * Integration (v2prov Framework)
    * Validation (Go Framework)
    * Other - Explain: _EXPLAIN_
    * None
    * _REMOVE NOT APPLICABLE BULLET POINTS ABOVE_
* If "None" - Reason: _EXPLAIN THE REASON_
  
* If "None" - GH Issue/PR: _LINK TO GH ISSUE/PR TO ADD TESTS_

Summary: _TODO_

### Regressions Considerations

_TODO_

Existing / newly added automated tests that provide evidence there are no regressions:
* _TODO_

### CODE REVIEW

File: pkg/controllers/management/agentupgrade/controller.go
Comment (bfbachmann) [inline pkg/controllers/management/agentupgrade/controller.go:64]: Minor: Might want to say something along the lines of "will return true if the pod is a cluster-agent or node-agent pod on version v2.0-v2.4 in the cattle-system namespace".... I _think_ that's what it's doing?

File: pkg/controllers/management/agentupgrade/controller.go
Comment (bfbachmann) [inline pkg/controllers/management/agentupgrade/controller.go:36]: Seems correct.

File: pkg/controllers/management/auth/crtb_handler.go
Comment (bfbachmann) [inline pkg/controllers/management/auth/crtb_handler.go:146]:
```go

// reconcileBindings translates a CRTB into several k8s roles and bindings when a CRTB is created or updated to actually enforce the RBAC
// Specifically:
//   - ensure the subject can see the cluster in the mgmt API
//   - if the subject was granted owner permissions for the cluster, ensure they can create/update/delete the cluster
//   - if the subject was granted privileges to mgmt plane resources that are scoped to the cluster, enforce those rules
//     in the cluster's mgmt plane namespace
```

File: pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go
Comment (bfbachmann) [inline pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go:609]: Is there a missing closing brace for the above function, or am I reading this wrong?

**Review by bfbachmann — COMMENTED** [review]
Wonderful! Thank you!

I left some very minor suggestions

File: pkg/controllers/management/agentupgrade/controller.go
Comment (a-blender) [reply pkg/controllers/management/agentupgrade/controller.go:64]: How version v2.0-v2.4?

File: pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go
Comment (a-blender) [reply pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go:612]: Accidental from the last set of merge conflicts. It's fixed and I made sure the repo builds successfully without errors

File: pkg/controllers/management/auth/prtb_handler.go
Comment (HarrisonWAffel) [inline pkg/controllers/management/auth/prtb_handler.go:241]: We should move this comment back into the body of the function, comments directly above a function should describe the function, this is more of a general note. This comment also applies to other functions in this package, like the reconcileBindings function

**Review by HarrisonWAffel — COMMENTED** [review]
just a couple of comments, haven't gone over every change yet

File: pkg/controllers/management/agentupgrade/controller.go
Comment (a-blender) [reply pkg/controllers/management/agentupgrade/controller.go:47]: Good idea! Based on the godocs, https://pkg.go.dev/github.com/fluhus/godoc-tricks#hdr-Documentation_Links [shouldDelete] should work in the comment but my local godoc server doesn't provide a reference for that. It will for package paths though. Not sure if this is an issue due to it being a local server or something

File: pkg/controllers/management/agentupgrade/controller.go
Comment (a-blender) [reply pkg/controllers/management/agentupgrade/controller.go:1]: I'm referencing the controller here, did you mean to request an update to the function description for `Register` to say it has two handlers in the controller?

File: pkg/controllers/management/agentupgrade/controller.go
Comment (HarrisonWAffel) [reply pkg/controllers/management/agentupgrade/controller.go:47]: Oh okay. I know that the IDE will provide a pointer when comments are written that way, which is useful when jumping around the code in Goland

Comment (a-blender) [reply pkg/controllers/management/agentupgrade/controller.go]: Yeah. I'll leave that reference as `shouldDelete` so it'll at least help devs auto jump to it in the code.

File: pkg/controllers/management/auth/project_cluster_handler.go
Comment (bfbachmann) [inline pkg/controllers/management/auth/project_cluster_handler.go:178]: Minor
```suggestion
// sync reconciles the state of a project for a cluster.
```

I think?

File: pkg/controllers/management/auth/project_cluster_handler.go
Comment (bfbachmann) [inline pkg/controllers/management/auth/project_cluster_handler.go:236]: Same here.

File: pkg/controllers/management/auth/project_cluster_handler.go
Comment (bfbachmann) [inline pkg/controllers/management/auth/project_cluster_handler.go:230]: Technically, this function does nothing because the create is handled by `sync`.

File: pkg/controllers/management/auth/project_cluster_handler.go
Comment (bfbachmann) [inline pkg/controllers/management/auth/project_cluster_handler.go:384]: Seems correct.
```suggestion
// ProjectRoleBinding if it doesn't exist to bind the RoleTemplate to the Project. It takes these actions
```

File: pkg/controllers/management/auth/user.go
Comment (bfbachmann) [inline pkg/controllers/management/auth/user.go:405]:
```suggestion
// deleteAllTokens deletes all of the given tokens.
```

File: pkg/controllers/management/auth/user.go
Comment (bfbachmann) [inline pkg/controllers/management/auth/user.go:457]: Minor:
```suggestion
// removeLegacyFinalizers removes the `controller.cattle.io/cat-user-controller` finalizer that from the given user.
```

**Review by bfbachmann — COMMENTED** [review]
Looks good for the most part, just minor stuff.

**Review by HarrisonWAffel — COMMENTED** [review]
Theres still a lot to cover here but I have a few more comments.

Because this is touching auth code we probably want to loop in team 1

**Review by bfbachmann — DISMISSED** [review]
LGTM! Thanks!

File: pkg/controllers/management/auth/manager.go
Comment (JonCrowther) [inline pkg/controllers/management/auth/manager.go:117]:
```suggestion
// painfully similar to ensureProjectMembershipBinding, but writing one function that handles both is overly complex.
```

File: pkg/controllers/management/auth/manager.go
Comment (JonCrowther) [inline pkg/controllers/management/auth/manager.go:306]:
```suggestion
// createMembershipRole creates a membership Role or ClusterRole for the given resource. Can specify if this is an owner role (ie give all verbs).
```

File: pkg/controllers/management/auth/prtb_handler.go
Comment (JonCrowther) [inline pkg/controllers/management/auth/prtb_handler.go:52]:
```suggestion
// Create ensures a ProjectRoleTemplateBinding (CRTB) has the correct Subject, Roles and RoleBindings on create.
```

File: pkg/controllers/management/auth/prtb_handler.go
Comment (JonCrowther) [inline pkg/controllers/management/auth/prtb_handler.go:65]:
```suggestion
// Update ensures a ProjectRoleTemplateBinding (CRTB) has the correct Subject, Labels, Roles and RoleBindings on update.
```

File: pkg/controllers/management/auth/prtb_handler.go
Comment (JonCrowther) [inline pkg/controllers/management/auth/prtb_handler.go:81]:
```suggestion
// Remove removes user access, privileges, and permissions for a PRTB.
```

File: pkg/controllers/management/auth/crtb_handler.go
Comment (JonCrowther) [inline pkg/controllers/management/auth/crtb_handler.go:223]: As mentioned here: https://github.com/rancher/rancher/pull/42473/files#r1339217868
This should not be moved out of the body

**Review by JonCrowther — CHANGES_REQUESTED** [review]

**Review by jakefhyde — CHANGES_REQUESTED** [review]

File: pkg/controllers/management/auth/project_cluster_handler.go
Comment (JonCrowther) [inline pkg/controllers/management/auth/project_cluster_handler.go:151]:
```suggestion
// Remove removes the set of namespaces and RoleBindings in the given project.
```

File: pkg/controllers/management/auth/project_cluster_handler.go
Comment (JonCrowther) [inline pkg/controllers/management/auth/project_cluster_handler.go:176]:
```suggestion
// sync reconciles the state of a Cluster in the management cluster.
```

File: pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go
Comment (a-blender) [reply pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go:612]: I have no idea, I'll remove it :)

File: pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go
Comment (a-blender) [reply pkg/controllers/management/auth/globalroles/globalrolebinding_handler.go:109]: The update funcs in the GlobalRole handler are all called `Updated` which isn't cohesive with the rest of the code. Should I update(pun intended) all those to `Update`?

**Review by bfbachmann — DISMISSED** [review]

**Review by jakefhyde — APPROVED** [review]

**Comment by a-blender** [comment]
Waiting on @JonCrowther re-review and approval to merge!

File: pkg/controllers/management/auth/manager.go
Comment (JonCrowther) [inline pkg/controllers/management/auth/manager.go:699]:
```suggestion
		// protect against a RoleBinding being in the list more than once (shouldn't happen, but just to be safe)
```

File: pkg/controllers/management/auth/manager.go
Comment (JonCrowther) [inline pkg/controllers/management/auth/manager.go:730]:
```suggestion
// checkForManagementPlaneRules returns the verbs from RoleTemplate rules that grant access to a management plane
```

File: pkg/controllers/management/auth/manager.go
Comment (JonCrowther) [inline pkg/controllers/management/auth/manager.go:296]:
```suggestion
// createProjectMembershipRole creates a Role that lets the bound subject see (if they are an ordinary member) the
```

File: pkg/controllers/management/auth/manager.go
Comment (JonCrowther) [inline pkg/controllers/management/auth/manager.go:306]:
```suggestion
// owner Role (ie give all verbs).
```

File: pkg/controllers/management/auth/manager.go
Comment (JonCrowther) [inline pkg/controllers/management/auth/manager.go:438]:
```suggestion
// removeAuthV2Permissions finds any RoleBindings based off the owner annotation from the incoming owner. This is
```

File: pkg/controllers/management/auth/manager.go
Comment (JonCrowther) [inline pkg/controllers/management/auth/manager.go:1]: We haven't been very consistent in capitalization of `Role` `RoleBinding` `RoleTemplate` etc. I personally prefer to capitalize it fully, but as you can see in this file some people don't. I'm ok with it either way, but in either case it should all be the same.

File: pkg/controllers/management/auth/manager.go
Comment (JonCrowther) [inline pkg/controllers/management/auth/manager.go:771]:
```suggestion
// reconcileManagementPlaneRole updates all Role rules for the management plane.
```

File: pkg/controllers/management/auth/manager.go
Comment (JonCrowther) [inline pkg/controllers/management/auth/manager.go:834]: 🆒

File: pkg/controllers/management/auth/manager.go
Comment (JonCrowther) [inline pkg/controllers/management/auth/manager.go:858]:
```suggestion
		// apiGroup. Checking for empty string also won't help since core api group is empty string.
```

File: pkg/controllers/management/auth/manager.go
Comment (JonCrowther) [inline pkg/controllers/management/auth/manager.go:851]:
```suggestion
// buildRule builds a policy rule with given resource and verbs.
```

File: pkg/controllers/management/auth/manager.go
Comment (JonCrowther) [inline pkg/controllers/management/auth/manager.go:873]:
```suggestion
// checkReferencedRoles checks referenced Roles for circular recursive calls, circular dependencies, and other
// reference issues.
```

**Review by JonCrowther — COMMENTED** [review]
Contents of the comments look good, just some small nits left.

**Comment by JonCrowther** [comment]
Out of curiosity, is there going to be any kind of linter added to CI going forward? Specifically for things like line length which I noticed you've updated. Ideally no one should have to make this huge effort again in the future.

**Comment by jakefhyde** [comment]
> Out of curiosity, is there going to be any kind of linter added to CI going forward? Specifically for things like line length which I noticed you've updated. Ideally no one should have to make this huge effort again in the future.

Just wanted to call out that it is specifically possible within `golangci-lint` to configure it such that specific set of rules are only run on net new code changes. We would still have to run the original as well, but we could become much more aggressive with linting moving forward.

**Review by JonCrowther — APPROVED** [review]

**Review by bfbachmann — DISMISSED** [review]

**Pull request closed without merging**
